examples/llm/src/models/layers/attention.py

```diff
@@ -12,377 +13,343 @@
-                                      self.d_model,
-                                      bias=True,
-                                      device=device)
-            # for param init fn
-            fuse_splits = (cfg.d_model, 2 * cfg.d_model)
-            self.Wqkv._fused = (0, fuse_splits)  # type: ignore
-            self.out_proj._is_residual = True  # type: ignore
-
-            if self.attn_qk_ln:
-                layernorm_class = nn.LayerNorm if not cfg.get(
-                    'low_precision_layernorm', False) else LPLayerNorm
-                self.q_ln = layernorm_class(self.d_model, device=device)
-                self.k_ln = layernorm_class(self.d_model, device=device)
-        else:
-            self.mhsa = FlashMHA(
-                embed_dim=cfg.d_model,
-                num_heads=cfg.n_heads,
-                bias=True,
-                batch_first=True,
-                causal=True,
-                softmax_scale=cfg.get('softmax_scale'),
-                device=device,
-            )
-            # for param init fn
-            fuse_splits = (cfg.d_model, 2 * cfg.d_model)
-            self.mhsa.Wqkv._fused = (0, fuse_splits)  # type: ignore
-            self.mhsa.out_proj._is_residual = True  # type: ignore
-
-        warnings.warn(
-            'While `attn_impl: triton` can be faster than `attn_impl: flash` '
-            'it uses more memory. When training larger models this can trigger '
-            'alloc retries which hurts performance. If encountered, we recommend '
-            'using `attn_impl: flash`.')
-
-    def forward(self, x, key_padding_mask=None, attn_mask=None):
-        if self.attn_qk_ln or self.clip_qkv:
-            qkv = self.Wqkv(x)
-            if self.clip_qkv:
-                qkv.clamp_(min=-self.clip_qkv, max=self.clip_qkv)
-            if self.attn_qk_ln:
-                # Applying layernorm to qk
-                dtype = qkv.dtype
-                q, k, v = qkv.split(self.d_model, dim=-1)
-                q = self.q_ln(q).to(dtype)
-                k = self.k_ln(k).to(dtype)
-                qkv = torch.cat([q, k, v], dim=-1)
-
-            # attention
-            context, attn_weights = self.inner_attn(
-                qkv,
-                key_padding_mask=key_padding_mask,
-                attn_mask=attn_mask,
-                is_causal=True)
-
-            return self.out_proj(context), attn_weights
+        self.out_proj._is_residual = True  # type: ignore
+
+    def forward(self,
+                x,
+                attn_bias=None,
+                key_padding_mask=None,
+                needs_weights=False):
+        qkv = self.Wqkv(x)
+        if self.clip_qkv:
+            qkv.clamp_(min=-self.clip_qkv, max=self.clip_qkv)
+        if self.attn_qk_ln:
+            # Applying layernorm to qk
+            dtype = qkv.dtype
+            q, k, v = qkv.split(self.d_model, dim=-1)
+            q = self.q_ln(q).to(dtype)
+            k = self.k_ln(k).to(dtype)
```

Comment:
are these `.to(dtype)` calls necessary? Is this to make sure that when using `nn.LayerNorm`, which upcasts to `fp32`, we immediately downcast the activations back to 16-bit?

Reply:
yes